File: src/main/java/io/serieznyi/intellij/factorioapicompletion/intellij/service/ApiService.java
package io.serieznyi.intellij.factorioapicompletion.intellij.service;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import lombok.CustomLog;
import io.serieznyi.intellij.factorioapicompletion.intellij.NotificationService;
import io.serieznyi.intellij.factorioapicompletion.core.parser.api.ApiParser;
import io.serieznyi.intellij.factorioapicompletion.core.version.ApiVersionCollection;
import io.serieznyi.intellij.factorioapicompletion.core.version.ApiVersionResolver;
import io.serieznyi.intellij.factorioapicompletion.core.version.FactorioApiVersion;
import io.serieznyi.intellij.factorioapicompletion.intellij.FactorioLibraryProvider;
import io.serieznyi.intellij.factorioapicompletion.intellij.FactorioState;
import io.serieznyi.intellij.factorioapicompletion.intellij.util.FilesystemUtil;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;

@CustomLog
public class ApiService {
    private final AtomicBoolean downloadInProgress = new AtomicBoolean(false);
    private final Project project;
    private final ApiParser apiParser;

    private ApiService(Project project) {
        this.project = project;

        Path pluginDir = FilesystemUtil.getPluginDir();
        Path apiRootPath = pluginDir.resolve("factorio_api");
        apiParser = new ApiParser(apiRootPath);
    }

    public static ApiService getInstance(Project project)
    {
        return new ApiService(project);
    }

    public void removeLibraryFiles() {
        if (downloadInProgress.get()) {
            return;
        }

        apiParser.removeCurrentAPI();
        FactorioLibraryProvider.reload();
    }

    public void checkForUpdate() {
        FactorioState config = FactorioState.getInstance(project);

        if (!config.useLatestVersion) {
            return;
        }

        FactorioApiVersion newestVersion = detectLatestAllowedVersion();

        if (newestVersion != null && !newestVersion.equals(config.selectedFactorioVersion)) {
            removeLibraryFiles();

            if (downloadInProgress.compareAndSet(false, true)) {
                ProgressManager.getInstance().run(new ApiTask());
            }
        }
    }

    public Optional<Path> getApiPath() {
        if (downloadInProgress.get()) {
            return Optional.empty();
        }

        FactorioApiVersion version = FactorioState.getInstance(project).selectedFactorioVersion;

        Optional<Path> path = apiParser.getApiPath(version);

        if (path.isEmpty() && downloadInProgress.compareAndSet(false, true)) {
            ProgressManager.getInstance().run(new ApiTask());
        }

        return path;
    }

    private FactorioApiVersion detectLatestAllowedVersion() {
        ApiVersionCollection factorioApiVersions;

        try {
            factorioApiVersions = (new ApiVersionResolver()).supportedVersions();
        } catch (Exception e) {
            log.error(e);
            NotificationService.getInstance(project).notifyErrorApiUpdating();
            return null;
        }

        return factorioApiVersions.latestVersion();
    }

    private class ApiTask extends Task.Backgroundable {
        public ApiTask() {
            super(project, "Download and Parse Factorio API", false);
        }

        @Override
        public void run(@NotNull ProgressIndicator indicator) {
            try {
                FactorioApiVersion selectedVersion = FactorioState.getInstance(project).selectedFactorioVersion;

                apiParser.parse(selectedVersion);

                ApplicationManager.getApplication().invokeLater(FactorioLibraryProvider::reload);
            } catch (Throwable e) {
                log.error(e);
                NotificationService.getInstance(project).notifyErrorApiUpdating();
            } finally {
                downloadInProgress.set(false);
            }
        }
    }
}

File: src/main/kotlin/io/serieznyi/intellij/factorioapicompletion/intellij/ApiService.kt
package io.serieznyi.intellij.factorioapicompletion.intellij

import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import io.serieznyi.intellij.factorioapicompletion.core.factorio.api.parser.ApiParser
import io.serieznyi.intellij.factorioapicompletion.core.factorio.api.source.FileCacheDataSourceFactory
import io.serieznyi.intellij.factorioapicompletion.core.factorio.api.source.HttpDataSourceFactory
import io.serieznyi.intellij.factorioapicompletion.core.factorio.writer.ApiWriterWrapper
import io.serieznyi.intellij.factorioapicompletion.core.factory.ApiVersionResolverFactory
import io.serieznyi.intellij.factorioapicompletion.core.factory.CacheFactory
import io.serieznyi.intellij.factorioapicompletion.core.util.io.findOrCreateDirectory
import java.nio.file.Path

@Service(Service.Level.PROJECT)
class ApiService(private val project: Project) {
    fun downloadApi() {
        val apiDir = FilesystemUtil.apiVersionsDir()
        val cache = CacheFactory.fileCache()
        val apiParser = ApiParser(FileCacheDataSourceFactory(HttpDataSourceFactory(), cache), false)
        val apiVersion = ApiVersionResolverFactory.create().supportedVersions().latestVersion()
        val writer = ApiWriterWrapper()
        val apiData = apiParser.parse(apiVersion)

        writer.writeAll(apiVersion, apiData, apiDir)

        this.triggerReindex()
    }

    fun triggerReindex() {
        FactorioApiProvider.reload(project)
    }

    fun apiDir(): Path {
        val settings = project.service<PluginSettings>()
        val pluginDir = FilesystemUtil.pluginDir()
        val apiDir = pluginDir.resolve("api").findOrCreateDirectory()

        return apiDir.resolve(settings.selectedApiVersion.version)
    }
}
